libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -51,251 +51,72 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
 
 	Type const* targetBaseType = _targetType.baseType();
 	Type const* sourceBaseType = _sourceType.baseType();
-
-	// TODO unroll loop for small sizes
-
-	bool sourceIsStorage = _sourceType.location() == DataLocation::Storage;
 	bool fromCalldata = _sourceType.location() == DataLocation::CallData;
-	bool directCopy = sourceIsStorage && sourceBaseType->isValueType() && *sourceBaseType == *targetBaseType;
-	bool haveByteOffsetSource = !directCopy && sourceIsStorage && sourceBaseType->storageBytes() <= 16;
-	bool haveByteOffsetTarget = !directCopy && targetBaseType->storageBytes() <= 16;
-	unsigned byteOffsetSize = (haveByteOffsetSource ? 1u : 0u) + (haveByteOffsetTarget ? 1u : 0u);
+	bool hasDynamicCalldataLength = fromCalldata && _sourceType.isDynamicallySized();
```

Comment:
I find having this condition inline much clearer than its current name. But if you really want to give it a name, I'd use one that conveys why we want to check it:

```suggestion
	bool haveSourceLengthOnStack = fromCalldata && _sourceType.isDynamicallySized();
```

Reply:
By the way, we can also remove `targetBaseType` and `sourceBaseType`. They made more sense in the past when the expressions were much longer. Now they're just an unnecessary indirection.

Reply:
Removed. But I kept the `sourceBaseType` since it does make sense to keep it I think. It is still being used in multiple places.

Reply:
I'd still remove it. It's not even significantly shorter than the value. I generally prefer avoiding unnecessary locals, because the added indirection hurts readability.